src/qspi.rs

```diff
@@ -0,0 +1,230 @@
+use crate::{
+    stm32,
+    time::Hertz,
+    rcc::Ccdr,
+};
+
+use core::ptr;
+
+#[derive(Debug)]
+pub enum QspiMode {
+    OneBit,
+    TwoBit,
+    FourBit,
+}
+
+#[derive(Debug)]
+pub enum QspiError {
+    Busy,
+    InvalidFrequency,
+    Unsupported,
+    InvalidAddress,
+    Underflow,
+    FifoData,
+    InvalidClock,
+}
+
+// TODO: Implement traits for valid QSPI pins and mark them as part of the structure for validation.
+pub struct Qspi {
+    rb: stm32::QUADSPI,
+}
+
+impl Qspi {
+    pub fn new<T>(regs: stm32::QUADSPI, ccdr: &Ccdr, frequency: T) -> Result<Self, QspiError>
+    where
+        T: Into<Hertz>,
+    {
+        // Enable quad SPI in the clocks.
+        ccdr.rb.ahb3enr.modify(|_, w| w.qspien().enabled());
+
+        let spi_kernel_ck = match Self::get_clock(&ccdr) {
+            Some(freq_hz) => freq_hz.0,
+            _ => return Err(QspiError::InvalidClock),
+        };
+
+        // Disable QUADSPI before configuring it.
+        regs.cr.write(|w| {
+            w.en().clear_bit()
+        });
+
+        // Configure the FSIZE to maximum. It appears that even when addressing is not used, the
+        // flash size violation may still trigger.
+        regs.dcr.write(|w| unsafe {
+            w.fsize().bits(0x1F)
+        });
+
+        // Clear all pending flags.
+        regs.fcr.write(|w| {
+            w
+            .ctof().set_bit()
+            .csmf().set_bit()
+            .ctcf().set_bit()
+            .ctef().set_bit()
+        });
+
+        // Configure the communication method for QSPI.
+        regs.ccr.write(|w| unsafe {
+            w
+            .fmode().bits(0)
+            .dmode().bits(0b01)
+            .imode().bits(0b01)
+        });
+
+        let spi_frequency = frequency.into().0;
+        let divisor = match spi_kernel_ck / spi_frequency {
+            1..=255 => {
+                spi_kernel_ck / spi_frequency - 1
+            },
+            _ => {
+                return Err(QspiError::InvalidFrequency);
+            },
+        };
+
+        // Write the prescaler and select flash bank 2 - flash bank 1 is currently unsupported.
+        regs.cr.write(|w| unsafe {
+            w.prescaler().bits(divisor as u8)
+             .fsel().set_bit()
+        });
+
```

Comment:
I think you want to `cr.en()` the peripheral here and leave it enabled. Don't toggle it during reads/writes: the pertinent registers can be changed with the peripheral enabled and transactions are triggered not by enabling but by writes to the various data/instruction/address registers depeding on the mode.

Reply:
After re-reading the documentation, I think you're right here. I'll remove the enables/disables outside of the constructor.